sql: TestDropDatabaseDeleteData failed

sql.TestDropDatabaseDeleteData [failed](https://mesolite.cluster.engflow.com/invocations/default/3d650e45-81ca-481b-89ab-3b418ab57e0c?testReportRun=1&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [5207cd59f4cfd8444cb3d7739f53063bed4ae1a6](https://github.com/cockroachdb/cockroach/commits/5207cd59f4cfd8444cb3d7739f53063bed4ae1a6):


```
=== RUN   TestDropDatabaseDeleteData
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestDropDatabaseDeleteData255376091
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    drop_test.go:342: condition failed to evaluate within 45s: from drop_test.go:95: descriptor exists expected=false was=true
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestDropDatabaseDeleteData255376091
--- FAIL: TestDropDatabaseDeleteData (45.90s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropDatabaseDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45955

sql.TestDropDatabaseDeleteData [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18484415?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18484415?buildTab=artifacts#/) on master @ [8adde2a89c0f8db7ba2aa8a946dbde2379b8f5b9](https://github.com/cockroachdb/cockroach/commits/8adde2a89c0f8db7ba2aa8a946dbde2379b8f5b9):


```
=== RUN   TestDropDatabaseDeleteData
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/f4d1e5a362e4384d18c9c26e2defb957/logTestDropDatabaseDeleteData3904146637
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:139: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    drop_test.go:342: condition failed to evaluate within 45s: from drop_test.go:95: descriptor exists expected=false was=true
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f4d1e5a362e4384d18c9c26e2defb957/logTestDropDatabaseDeleteData3904146637
--- FAIL: TestDropDatabaseDeleteData (45.97s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropDatabaseDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestDropDatabaseDeleteData [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18576395?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18576395?buildTab=artifacts#/) on master @ [8b1e4c9ed7f0b7b7514571a139d122229df5d3ee](https://github.com/cockroachdb/cockroach/commits/8b1e4c9ed7f0b7b7514571a139d122229df5d3ee):


```
=== RUN   TestDropDatabaseDeleteData
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/37acaf1de75f910fa45ac01dbd64243a/logTestDropDatabaseDeleteData3337253929
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    drop_test.go:342: condition failed to evaluate within 45s: from drop_test.go:95: descriptor exists expected=false was=true
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/37acaf1de75f910fa45ac01dbd64243a/logTestDropDatabaseDeleteData3337253929
--- FAIL: TestDropDatabaseDeleteData (47.93s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139496 sql: TestDropDatabaseDeleteData failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropDatabaseDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestDropDatabaseDeleteData [failed](https://tanzanite.cluster.engflow.com/invocations/default/854fece1-32a8-409a-89ae-58887ab835de?testReportRun=7&testReportShard=10&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [3acb4cd1a4369c93718975681e228ffd74007832](https://github.com/cockroachdb/cockroach/commits/3acb4cd1a4369c93718975681e228ffd74007832):


```
=== RUN   TestDropDatabaseDeleteData
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestDropDatabaseDeleteData2590405882
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    drop_test.go:342: condition failed to evaluate within 45s: from drop_test.go:95: descriptor exists expected=false was=true
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestDropDatabaseDeleteData2590405882
--- FAIL: TestDropDatabaseDeleteData (45.72s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=10</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139496 sql: TestDropDatabaseDeleteData failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropDatabaseDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestDropDatabaseDeleteData [failed](https://tanzanite.cluster.engflow.com/invocations/default/9d17ad83-65dd-4b6e-a688-e4453de0cb36?testReportRun=2&testReportShard=10&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [cc3365feadf12f4669a164f8033ad3ddf1f80fc3](https://github.com/cockroachdb/cockroach/commits/cc3365feadf12f4669a164f8033ad3ddf1f80fc3):


```
=== RUN   TestDropDatabaseDeleteData
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestDropDatabaseDeleteData3106349216
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    drop_test.go:351: expecting 6 keys, but found 0
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestDropDatabaseDeleteData3106349216
--- FAIL: TestDropDatabaseDeleteData (16.74s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=10</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139496 sql: TestDropDatabaseDeleteData failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropDatabaseDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
+++ b/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
@@ -298,6 +298,11 @@ func (s *SQLWatcher) watchForZoneConfigUpdates(
 	}
 
 	log.Infof(ctx, "established range feed over system.zones starting at time %s", startTS)
+
+	if s.knobs != nil && s.knobs.OnWatchForZoneConfigUpdatesEstablished != nil {
+		s.knobs.OnWatchForZoneConfigUpdatesEstablished()
+	}
+
 	return rf, nil
 }
 

--- a/pkg/spanconfig/testing_knobs.go
+++ b/pkg/spanconfig/testing_knobs.go
@@ -126,6 +126,10 @@ type TestingKnobs struct {
 	// OverrideFallbackConf, if set, allows tests to override fields in the
 	// fallback config that will be applied to the span.
 	OverrideFallbackConf func(roachpb.SpanConfig) roachpb.SpanConfig
+
+	// OnWatchForZoneConfigUpdatesEstablished is invoked when the RangeFeed over
+	// system.zones starts.
+	OnWatchForZoneConfigUpdatesEstablished func()
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

--- a/pkg/sql/drop_test.go
+++ b/pkg/sql/drop_test.go
@@ -10,6 +10,7 @@ import (
 	gosql "database/sql"
 	"fmt"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -23,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkeys"
@@ -265,6 +267,22 @@ func TestDropDatabaseDeleteData(t *testing.T) {
 	// Speed up mvcc queue scan.
 	params.ScanMaxIdleTime = time.Millisecond
 
+	// zoneCfgRangeFeedStarted is used to signal that the zone config range feed
+	// started. This is used to avoid a race where we update `system.zones` before
+	// the full reconciliation of zone configs has started. The full
+	// reconciliation ignores dropped databases, and if we write to
+	// `system.zones` before that, our write will not be reconciled. By delaying
+	// the test until the zone config range feed starts, we ensure that the full
+	// reconciliation has finished, and the range feed is ready to stream our
+	// changes.
+	var zoneCfgRangeFeedStarted sync.WaitGroup
+	zoneCfgRangeFeedStarted.Add(1)
+	params.Knobs.SpanConfig = &spanconfig.TestingKnobs{
+		OnWatchForZoneConfigUpdatesEstablished: func() {
+			zoneCfgRangeFeedStarted.Done()
+		},
+	}
+
 	srv, sqlDB, kvDB := serverutils.StartServer(t, params)
 	defer srv.Stopper().Stop(context.Background())
 	ctx := context.Background()
@@ -273,11 +291,23 @@ func TestDropDatabaseDeleteData(t *testing.T) {
 	systemDB := srv.SystemLayer().SQLConn(t)
 	_, err := systemDB.Exec(`SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';`)
 	require.NoError(t, err)
-
 	// Refresh protected timestamp cache immediately to make MVCC GC queue to
 	// process GC immediately.
 	_, err = systemDB.Exec(`SET CLUSTER SETTING kv.protectedts.poll_interval = '1s';`)
 	require.NoError(t, err)
+
+	// Turn off storage_coalesce_adjacent to avoid a race between the GC and the
+	// split queues where a range might get GCed before splitting. This is
+	// more likely to happen in this test since increase the GC queue scan rate.
+	// If we don't turn the setting off, setting the TTL to table 1 might cause
+	// table 2 to also get purged.
+	_, err = systemDB.Exec(`SET CLUSTER SETTING spanconfig.storage_coalesce_adjacent.enabled = 'false';`)
+	require.NoError(t, err)
+	_, err = systemDB.Exec(`SET CLUSTER SETTING spanconfig.tenant_coalesce_adjacent.enabled = 'false';`)
+	require.NoError(t, err)
+
+	// Wait for the zone config range feed to start.
+	zoneCfgRangeFeedStarted.Wait()
 
 	// Disable strict GC TTL enforcement because we're going to shove a zero-value
 	// TTL into the system with AddImmediateGCZoneConfig.

